Microsoft.Toolkit.HighPerformance/Buffers/StringPool.cs

```diff
@@ -0,0 +1,407 @@
+                    {
+                        return entry;
+                    }
+
+                    // ReadOnlySpan<char>.ToString() creates a string with the span contents.
+                    // This is equivalent to doing new string(span) on Span<T>-enabled runtimes,
+                    // or to using an unsafe block, a fixed statement and new string(char*, int, int).
+                    return entry = span.ToString();
+                }
+            }
+
+            /// <summary>
+            /// Implements <see cref="StringPool.TryGet"/> for the current <see cref="Bucket"/> instance.
+            /// </summary>
+            /// <param name="span">The input <see cref="ReadOnlySpan{T}"/> with the contents to use.</param>
+            /// <param name="hashcode">The precomputed hashcode for <paramref name="span"/>.</param>
+            /// <param name="value">The resulting cached <see cref="string"/> instance, if present</param>
+            /// <returns>Whether or not the target <see cref="string"/> instance was found.</returns>
+            public bool TryGet(ReadOnlySpan<char> span, int hashcode, [NotNullWhen(true)] out string? value)
+            {
+                lock (this.dummy)
+                {
+                    ref string?[]? entries = ref this.entries;
+
+                    if (!(entries is null))
+                    {
+                        int entryIndex = hashcode % this.entriesPerBucket;
+
+                        ref string? entry = ref entries.DangerousGetReferenceAt(entryIndex);
+
+                        if (!(entry is null) &&
+                            entry.AsSpan().SequenceEqual(span))
+                        {
+                            value = entry;
+
+                            return true;
+                        }
+                    }
+
+                    value = null;
+
+                    return false;
+                }
+            }
+
+            /// <summary>
+            /// Resets the current array of entries.
+            /// </summary>
+            public void Reset()
+            {
+                lock (this.dummy)
+                {
+                    this.entries = null;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Gets the (positive) hashcode for a given <see cref="ReadOnlySpan{T}"/> instance.
+        /// </summary>
+        /// <param name="span">The input <see cref="ReadOnlySpan{T}"/> instance.</param>
+        /// <returns>The hashcode for <paramref name="span"/>.</returns>
+        [Pure]
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static int GetHashCode(ReadOnlySpan<char> span)
+        {
+            // We calculate the content hashcode for the input span and
+            // perform a XOR with the input length, to try to reduce collisions
+            // in case two sequences of different length result in the same one.
+            return
+                span.Length ^
```

Comment:
I am inclined to question whether XOR'ing with length is beneficial here

Reply:
I was looking at this and it seemed to me that especially for smaller inputs (say, words smaller than 10-15 characters), a lot of the collisions were between words of different lengths - I figured combining the length info as well could occasionally help? I mean, I don't actually have data on this so it might very well not be that effective in practice (I just imagine it wouldn't hurt?). Do you think we should just stick to the hashcode on its own? 🤔

Reply:
I just feel like it probably is lower quality entropy than what DJB2 is doing so my gut says it isn't a good idea. Won't block on this tho

Reply:
Yeah that's a fair point. Fixed this in https://github.com/windows-toolkit/WindowsCommunityToolkit/pull/3380/commits/14896611b4c2e6133b02d201e7a15df083938b75 👍